### ISSUE TITLE
Multiple Threads using DataLoader

### ISSUE THREAD

**Issue by dBug404** [graphql-dotnet/graphql-dotnet#795]
I recently migrated to GraphQL NET 2.0.0. 

When I implemented the dataloader in a query it worked fine, but in in more complex queries I encountered, that some other resolve methods are executed in a different thread and HttpContext.Current is null which causes problems. This disappeared, when I refactored the resolve with the dataloader back to not using dataloader.

The query is executed in a WebAPI controller in .NET 4.7.

Is this behaviour intended or is this a bug? Or can I configure, if the resolution stays in the same context so that I can use HttpContext.Current?

**Comment by dBug404** [comment]
You are right and in new code this is done. But in my case GraphQL is used on top of a bigger legacy backend so that is not really an option in the near future.

In most cases HttpContext.Current is abstracted via interfaces. So I could pass the HttpContext.Current via UserContext to the resolve methods and then override the standard DI registration for each specific DI resolution in the resolve method. 

But in my opinion it would be preferable, if the user can configure if multiple threads are used or not.

**Comment by johnrutherford** [comment]
The DataLoader doesn’t do anything with threads. There are unit tests that run within a synchronization context to make sure there are no issues with it. 

Is it possible that your code is creating threads or queuing work on the `ThreadPool` with `Task.Run()` or something? 

Could you create a simple app that reproduces the issue?

**Comment by dBug404** [comment]
I've created a simple app https://github.com/dBug404/graphql-dotnet-issue-795

Both projects can reproduce the issue. In graphql.web.simple I removed all StructureMap DI code.

The endpoint url should be: http://localhost:8153/api/graphql

There is a queries.txt file which contains the sample query with and without debug information.

![image](https://user-images.githubusercontent.com/2844552/44657703-08d89880-a9fe-11e8-91a8-b5f3b13562e7.png)


The only place where the code does anything with Tasks is the func in the DataLoaderContextAccessor where I use Task.FromResult(...) to support non async methods in the backend. But this should not be responsible for starting new Threads, right?

`

            Field<DealerGraphType>("dealer", resolve:
                context =>
                {
                    //return dealerService.GetDealer(context.Source.DealerId);
                    var loader = new DataLoaderContextAccessor().Context.GetOrAddBatchLoader<string, Dealer>("getDealerByDealerId", dealerIds =>
                    {
                        IDictionary<string, Dealer> dictionary = new DealerService()
                            .GetDealers(dealerIds.ToList())
                            .ToDictionary(dealer => dealer.DealerId);

                        return Task.FromResult(dictionary);
                    });
                    return loader.LoadAsync(context.Source.DealerId);
                });
`

**Comment by johnrutherford** [comment]
@dBug404, thanks. I've looked into it a bit more. It looks like it's happening because we're using `ConfigureAwait(false)` in a number of places. Continuations aren't run in the same synchronization context and so `HttpContext.Current` is `null`. If we removed `ConfigureAwait(false)` everywhere, it's possible to cause deadlocks for other users depending on how they are calling things. 

The DataLoader code doesn't explicitly create threads or queue task on the `ThreadPool` but it does do a lot of async stuff. And that's necessary to be able to batch requests. 

So I agree with @joemcbride, that it would be best to pass your `HttpContext` instance using the `UserContext`.
